### PR TITLE
Implement crate publishing on CI

### LINKED ISSUE #12678: Annotate crates not meant to be published with `publish = false`
For the upcoming Automated Substrate crates' Release effort, it'll be easier to detect if a crate is not meant to be published if its `Cargo.toml` includes a `publish = false`. This is recommended in https://doc.rust-lang.org/cargo/reference/manifest.html#the-publish-field and also is a common practice for other Rust projects, e.g. https://github.com/awslabs/smithy-rs/blob/b2c989630fe42560264be8fe917f3f1950c8e30f/rust-runtime/aws-smithy-types/fuzz/Cargo.toml#L5.

The absence of those annotations isn't a blocker for the publishing since we can work around it through custom exclusion lists, but it'd be nice to have those annotations directly in the repository.

In case all crates are meant to be published, feel free to close this ticket.

### PR DESCRIPTION
**Pull request by joao-paulo-parity**
related to https://github.com/paritytech/release-engineering/issues/129 and https://github.com/paritytech/release-engineering/issues/122

See https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2107373 for an example.

Note that, although the publishing job is set up for `master`, publishing to crates.io won't be enabled quite yet because some parts of the publishing script are disabled for now (https://github.com/paritytech/releng-scripts/pull/20). The idea is to try the script out for some time while only publishing to the local instance, and after we see that it works fine, the parts of the script which publishes to crates.io will be re-enabled gradually. We also need to resolve https://github.com/paritytech/release-engineering/issues/132 before that.

Some notes for reviewers

- The publishing automation strips away `[dev-dependencies]` before publishing crates, thus some dependencies have been copied from `[dev-dependencies]` to `[dependencies]` as **optional** entries so that `cargo` can find them while verifying the packages' features before publishing.
    Note also that their `[dependencies]` entries don't include a `version` field in their import declaration because it's not relevant for them.
- The CI job's time to completion should be improved substantially after `rusty-cachier` kicks in. `rusty-cachier` 's cache will only be created after the job runs on `master`.

close #12678
close https://github.com/paritytech/release-engineering/issues/122
close https://github.com/paritytech/release-engineering/issues/129

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2104872

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2104914

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2104915

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2104930

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2104975

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2104977

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2104958

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2104959

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2104965

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2104964

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2104970

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2105008

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2105117

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2105270

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2105269

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2105313

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2105368

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2105367

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2105369

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2105413

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2105415

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2105414

**Comment by joao-paulo-parity** [comment]
FYI the CI failures of https://gitlab.parity.io/parity/mirrors/substrate/builds/2107936 and https://gitlab.parity.io/parity/mirrors/substrate/builds/2107935 have nothing to do with the inclusions in this PR

File: scripts/ci/gitlab/pipeline/publish.yml [hunk 1]
```diff
@@ -161,3 +161,36 @@ update-node-template:
...
+  stage:                           publish
+  extends:                         .crates-publishing-template
+  # We don't want multiple jobs racing to publish crates as it's redundant and they might overwrite
+  # the releases of one another. Use resource_group to ensure that at most one instance of this job
+  # is running at any given time.
+  resource_group:                  crates-publishing
+  variables:
+    # crates.io rate limits crates publishing by 1 per minute, so a delay needs to be inserted
+    # slightly higher than that after publishing each crate. The value is specified in seconds.
+    SPUB_AFTER_PUBLISH_DELAY: 64
```
Comment (joao-paulo-parity) [inline scripts/ci/gitlab/pipeline/publish.yml:175]: https://github.com/paritytech/release-engineering/issues/123#issuecomment-1335509748

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2114926

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2114933

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2114927

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2114928

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2114922

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2114921

**Review by alvicsam — APPROVED** [review]

**Review by gilescope — APPROVED** [review]
It's going to make docs.rs relevant again so I'm in favour.

**Review by sergejparity — APPROVED** [review]

**Comment by joao-paulo-parity** [comment]
bot merge

**Comment by bkchr** [comment]
> * The publishing automation strips away `[dev-dependencies]` before publishing crates, thus some dependencies have been copied from `[dev-dependencies]` to `[dependencies]` as **optional** entries so that `cargo` can find them while verifying the packages' features before publishing.
>   Note also that their `[dependencies]` entries don't include a `version` field in their import declaration because it's not relevant for them.

Okay this is completely wrong. I would be very happy that we don't do these things nor that we approve these kind of things. The underlying issue here being that we enable features for dev dependencies, which is wrong. To solve this issue properly, you need to remove the enabling of features from dev dependencies.

Next time please reach out to the devs before doing this.

**Comment by bkchr** [comment]
This fixes the crates: https://github.com/paritytech/substrate/pull/12899

**Comment by joao-paulo-parity** [comment]
> Okay this is completely wrong. I would be very happy that we don't do these things nor that we approve these kind of things. The underlying issue here being that we enable features for dev dependencies, which is wrong. To solve this issue properly, you need to remove the enabling of features from dev dependencies.

The change was made to work around how the code was set up at the time. I couldn't intuit that the code was "wrong" before since it was presumably approved by the devs the way it was. Glad to know that there is a proper fix in https://github.com/paritytech/substrate/pull/12899.

**Comment by bkchr** [comment]
Yes I know that the initial "wrong" cargo tomls weren't added by you. However, it failing in your process you could have asked in Substrate channel or similar on these things!

**Comment by bkchr** [comment]
Thank you @joao-paulo-parity :)

**Pull request merged**